### ISSUE TITLE
Supporting for `<select>` and `<selectedoption>`

### ISSUE THREAD

**Issue by YusukeHirao** [markuplint/markuplint#2069]
## Document or Issue Link

- https://github.com/whatwg/html/issues/9799
- https://github.com/whatwg/html/pull/10633

**Comment by YusukeHirao** [comment]
## parse5 の対応状況 (2026-02-09)

whatwg/html#10548 は 2025-07-22 にマージ済みですが、parse5 側の対応はまだ完了していません：

- inikulin/parse5#1317 — "Relax `<select>` parser (WHATWG proposal)" → **DRAFT**
- inikulin/parse5#1610 — "Implement `<select>` parsing changes" → **OPEN**（サブモジュール更新のみ）

`<select>` 内の `<button>`, `<div>`, `<selectedcontent>` 等のパースが正しく行われないため、`@markuplint/html-spec` への `<selectedcontent>` 要素追加と `<select>` コンテンツモデル拡張は parse5 の対応後に行う予定です。

### 対応予定の仕様変更

- `spec.selectedcontent.json` 新規追加（implicitRole: `generic`, permittedRoles: any, namingProhibited, contentModel: nothing）
- `spec.select.json` コンテンツモデル拡張（`button`, `selectedcontent`, `div`, `noscript` を追加）
- 関連: `<optgroup>` に `legend` 子要素、`<option>` コンテンツモデル拡張、`<div>` カテゴリ追加

**Comment by YusukeHirao** [comment]
The nu-html-checker compatibility benchmark (#3582) reveals ~5 false positives related to the new customizable `<select>` content model:

- `html/elements/select/button-child-isvalid.html` — text node not allowed in `<select>`
- `html/elements/select/button-in-multiple-size1-isvalid.html` — text node not allowed in `<select>`
- `html/elements/select/button-selectedcontent-isvalid.html` — text node not allowed in `<select>`
- `html/elements/optgroup/div-child-isvalid.html` — text node not allowed in `<optgroup>`
- `html/elements/optgroup/legend-child-isvalid.html` — text node not allowed in `<optgroup>`

These are valid under the new `<select>` spec but markuplint's current `permitted-contents` model rejects them.

**Comment by YusukeHirao** [comment]
## Update (2026-08-30): still blocked on parse5, but here's a concrete plan for when it lands

### Blocker status

`inikulin/parse5#1610` ("Implement `<select>` parsing changes") is still **OPEN**, not merged, as of 2026-08-30 (last activity 2026-05-04). The draft `#1317` is stale (last updated 2024-11-01). So the parser-level blocker mentioned in my 2026-02-09 comment above hasn't changed.

### Confirmed content models (re-verified against the live HTML Standard)

- `<select>`: "Zero or one `button` elements if the `select` is a drop-down box, followed by zero or more `option`, `optgroup`, `hr`, script-supporting, `noscript`, or `div` elements." Landed via whatwg/html#12504 (2026-06-04, "Make select/option content model clearer about children versus descendants") and whatwg/html#12568 (2026-07-07, "Remove categories of select/optgroup/option element inner content elements").
- `<optgroup>`: "Zero or one `legend` element followed by zero or more `option`, script-supporting, `noscript`, or `div` elements."
- `<option>`: without `label`/`value`, and not a descendant of `<datalist>`: "Zero or more `div` elements or phrasing content" (no interactive content descendant). Other combinations of `label`/`value`/`datalist` are unchanged from before.
- `<selectedcontent>`: phrasing content, no interactive content descendant. Allowed only as a child of `<option>`, or as the sole child of the first `<button>` inside `<select>` (max one).
- Per HTML-AAM (w3c.github.io/html-aria): `<selectedcontent>` has implicit role `generic`; both it and the first `<button>` child of `<select>` must have **no `role` or `aria-*` attributes** when used in that structural position.

### Re: #3591 / #3637 — likely closed prematurely

I re-verified locally against the current `dev` build. Both issues were closed on the basis of "the benchmark already shows `match-error`," but that's not the same as "the fix landed":

- `src/spec.optgroup.jsonc`, `src/spec.select.jsonc`, `src/spec.option.jsonc` still carry the pre-customizable-select content models — no `button`/`legend`/`div`/`noscript` additions.
- `src/spec.button.jsonc` has no ARIA condition for "first child of `<select>`".
- Running the fixtures locally, `permitted-contents` does flag them — but only via its generic "unexpected text node" check, which fires regardless of whether the surrounding element/attribute logic is correct. That's not evidence the intended content-model/ARIA data was ever updated.

I'd suggest reopening both once the actual spec-data fix lands, so the fix can be verified end-to-end instead of relying on the benchmark's pass/fail as a proxy.

### Sketch of the `src/*.jsonc` changes (for whenever parse5 unblocks this)

Not proposing these as final PRs yet — posting the shape of the fix so the next pass doesn't start from scratch, and flagging the open questions inline.

**`spec.select.jsonc`**
```jsonc
"contentModel": {
	"contents": [
		{ "optional": "button" },
		{ "zeroOrMore": ["option", "optgroup", "hr", "noscript", "div", "#script-supporting"] }
	]
}
```
Open question: the spec limits `button` to drop-down boxes (no `multiple`, `size` ≤ 1). The current content model doesn't branch on `multiple`/`size` at all, so the simplest fix allows `button` unconditionally and leaves the drop-down-only constraint to a separate ARIA/attribute rule — needs a decision.

**`spec.optgroup.jsonc`**
```jsonc
"contentModel": {
	"contents": [
		{ "optional": "legend" },
		{ "zeroOrMore": ["option", "noscript", "div", "#script-supporting"] }
	]
}
```

**`spec.option.jsonc`** (default `contents`, i.e. no `label`, not under `<datalist>`)
```jsonc
"contents": [
	{ "zeroOrMore": ["div", "#phrasing"] }
]
```
Replaces the current `{"require": "#nonEmptyText"}`. Open questions: (1) how to express "no interactive content descendant" in the content-model schema — `spec.button.jsonc`'s `":model(phrasing):not(:model(interactive), :has(:model(interactive)))"` selector pattern might be reusable; (2) dropping the non-empty-text requirement needs a check against the nu-validator benchmark for regressions.

**`spec.selectedcontent.jsonc`** (new file)
```jsonc
// https://html.spec.whatwg.org/multipage/form-elements.html#the-selectedcontent-element
// https://w3c.github.io/html-aria/#el-selectedcontent
{
	"contentModel": {
		"contents": [
			{ "zeroOrMore": ":model(phrasing):not(:model(interactive), :has(:model(interactive)))" }
		]
	},
	"globalAttrs": {
		"#HTMLGlobalAttrs": true,
		"#GlobalEventAttrs": true,
		"#ARIAAttrs": true
	},
	"attributes": {},
	"aria": {
		"implicitRole": "generic",
		"permittedRoles": true,
		"conditions": {
			":is(option, select > button:first-child) > selectedcontent": {
				"implicitRole": "generic",
				"permittedRoles": false,
				"properties": { "global": false, "role": false }
			}
		}
	}
}
```
`option`'s and `select > button`'s content models also need `selectedcontent` added to their permitted contents. Also needs a category entry in `spec-common.contents.jsonc` per the html-spec CLAUDE.md rule.

**`spec.button.jsonc`** — new `conditions` entry, same pattern as the existing `[popovertarget]` condition:
```jsonc
"select > button:first-child": {
	"implicitRole": false,
	"permittedRoles": false,
	"properties": { "global": false, "role": false }
}
```

Caveat: I haven't verified that a selector like `select > button:first-child` is actually supported by the content-model/ARIA condition matcher in `@markuplint/ml-spec` (`get-content-model.ts` and friends) — that needs checking against the actual condition-matching implementation before landing any of this.
